lib/registry/src/lib.rs

```diff
@@ -643,77 +642,94 @@ pub fn get_global_install_dir(
+            })
         } else {
-            ar.unpack(target_path)
-                .map_err(|e| format!("failed to unpack {}: {e}", target_targz_path.display()))
+            ar.unpack(target_path).map_err(|e| {
+                anyhow::anyhow!("failed to unpack {}: {e}", target_targz_path.display())
+            })
         }
     };
 
     let try_decode_xz = || {
         let file = open_file()?;
         let mut decomp: Vec<u8> = Vec::new();
         let mut bufread = std::io::BufReader::new(&file);
-        lzma_rs::xz_decompress(&mut bufread, &mut decomp)
-            .map_err(|e| format!("failed to unpack {}: {e}", target_targz_path.display()))?;
+        lzma_rs::xz_decompress(&mut bufread, &mut decomp).map_err(|e| {
+            anyhow::anyhow!("failed to unpack {}: {e}", target_targz_path.display())
+        })?;
 
         let cursor = std::io::Cursor::new(decomp);
         let mut ar = tar::Archive::new(cursor);
         if strip_toplevel {
-            unpack_sans_parent(ar, target_path)
-                .map_err(|e| format!("failed to unpack {}: {e}", target_targz_path.display()))
+            unpack_sans_parent(ar, target_path).map_err(|e| {
+                anyhow::anyhow!("failed to unpack {}: {e}", target_targz_path.display())
+            })
         } else {
-            ar.unpack(target_path)
-                .map_err(|e| format!("failed to unpack {}: {e}", target_targz_path.display()))
+            ar.unpack(target_path).map_err(|e| {
+                anyhow::anyhow!("failed to unpack {}: {e}", target_targz_path.display())
+            })
         }
     };
 
     try_decode_gz().or_else(|_| try_decode_xz())?;
 
-    let _ = std::fs::remove_file(target_targz_path);
+    Ok(target_targz_path.to_path_buf())
+}
+
+/// Whether the top-level directory should be stripped
+pub fn download_and_unpack_targz(
+    url: &str,
+    target_path: &Path,
+    strip_toplevel: bool,
+) -> Result<PathBuf, anyhow::Error> {
+    let tempdir = tempdir::TempDir::new("wasmer-download-targz")?;
+
+    let target_targz_path = tempdir.path().join("package.tar.gz");
+
+    let mut resp = reqwest::blocking::get(url)
+        .map_err(|e| anyhow::anyhow!("failed to download {url}: {e}"))?;
+
+    if !target_targz_path.exists() {
+        // create all the parent paths, only remove the created directory, not the parent dirs
+        let _ = std::fs::create_dir_all(&target_targz_path);
+        let _ = std::fs::remove_dir(&target_targz_path);
+    }
+
+    {
+        let mut file = std::fs::File::create(&target_targz_path).map_err(|e| {
+            anyhow::anyhow!(
+                "failed to download {url} into {}: {e}",
+                target_targz_path.display()
+            )
+        })?;
+
+        resp.copy_to(&mut file)
+            .map_err(|e| anyhow::anyhow!("{e}"))?;
+    }
+
+    try_unpack_targz(target_targz_path.as_path(), target_path, strip_toplevel)
+        .context(anyhow::anyhow!("Could not download {url}"))?;
```

Comment:
I think you meant `.with_context(|| anyhow::anyhow!("Could not download {url}"))` here. 

The `context()` method accepts a value, which means `anyhow::anyhow!()` will be capturing backtraces and creating errors every time we reach this line. Even if there were no errors.